src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -186,10 +184,57 @@ import scala.language.higherKinds
   *  @define orderDependentFold
   */
 @SerialVersionUID(3L)
-sealed abstract class LazyList[+A] extends AbstractSeq[A] with LinearSeq[A] with LazyListOps[A, LazyList, LazyList[A]] {
-  override def iterableFactory: LazyListFactory[LazyList] = LazyList
+final class LazyList[+A] private(private[this] var lazyState: () => LazyList.State[A])
+  extends AbstractSeq[A] with LinearSeq[A] with LinearSeqOps[A, LazyList, LazyList[A]] {
+  import LazyList._
+
+  @volatile private[this] var stateEvaluated: Boolean = false
+  @inline private def stateDefined: Boolean = stateEvaluated
+
+  private lazy val state: State[A] = {
+    stateEvaluated = true
+    val res = lazyState()
```

Comment:
Should the two lines above be swapped (in case evaluating the state throws)?

Reply:
I don't think so. `stateEvaluated` is an indicator of whether or not we can touch `state` without breaking laziness. IMO, even if evaluating `state` throws, it should be considered evaluated and safe to touch.